sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/DynamicDestinations.java

```diff
@@ -0,0 +1,159 @@
+import org.apache.beam.sdk.transforms.DoFn.ProcessContext;
+import org.apache.beam.sdk.values.PCollectionView;
+import org.apache.beam.sdk.values.ValueInSingleWindow;
+
+/**
+ * This class provides the most general way of specifying dynamic BigQuery table destinations.
+ * Destinations can be extracted from the input element, and stored as a custom type. Mappings
+ * are provided to convert the destination into a BigQuery table reference and a BigQuery schema.
+ * The class can read a side input from another PCollection while performing these mappings.
+ *
+ * <p>For example, consider a PCollection of events, each containing a user-id field. You want to
+ * write each user's events to a separate table with a separate schema per user. Since the user-id
+ * field is a string, you will represent the destination as a string.
+ * <pre>{@code
+ * events.apply(BigQueryIO.<UserEvent>write()
+ *  .to(new DynamicDestinations<UserEvent, String>() {
+ *        public String getDestination(ValueInSingleWindow<String> element) {
+ *          return element.getValue().getUserId();
+ *        }
+ *        public TableDestination getTable(String user) {
+ *          return new TableDestination(tableForUser(user), "Table for user " + user);
+ *        }
+ *        public TableSchema getSchema(String user) {
+ *          return tableSchemaForUser(user);
+ *        }
+ *      })
+ *  .withFormatFunction(new SerializableFunction<UserEvent, TableRow>() {
+ *     public TableRow apply(UserEvent event) {
+ *       return convertUserEventToTableRow(event);
+ *     }
+ *   }));
+ * }</pre>
+ *
+ * <p>An instance of {@link DynamicDestinations} can also request a side input value that can be
+ *  examined from inside {@link #getTable} and {@link #getSchema}. It must return a list of the
+ *  side inputs it needs access to in {@link #getSideInputs()}. The value can be examined via the
+ *  {@link DynamicDestinations.SideInputAccessor} parameter passed into {@link #getTable}
+ *  and {@link #getSchema}.
+ *
+ * <p>{@code DestinationT} is expected to provide proper hash and equality members. Ideally it will
+ * be a compact type with an efficient coder, as these objects may be used as a key in a
+ * {@link org.apache.beam.sdk.transforms.GroupByKey}.
+ */
+public abstract class DynamicDestinations<T, DestinationT> implements Serializable {
+  /**
+   * Returns the materialized value of the side input. Can be used by concrete
+   * {@link DynamicDestinations} instances in {@link #getSchema} or {@link #getTable}.
+   */
+  public static class SideInputAccessor {
+    private ProcessContext processContext;
+    SideInputAccessor(ProcessContext processContext) {
+      this.processContext = processContext;
+    }
+
+    public <SideInputT> SideInputT getSideInputValue(PCollectionView<SideInputT> sideInput)
+        throws IllegalStateException {
+      @SuppressWarnings("unchecked")
+      SideInputT materialized = (SideInputT)  processContext.sideInput(sideInput);
```

Comment:
I believe this cast and SuppressWarnings is unnecessary. Simply say "return processContext.sideInput(sideInput)"

I guess it might be complaining cause your ProcessContext is a raw type - in that case, consider changing its type to be `DoFn<?, ?>.ProcessContext`.